### PR TITLE
[@types/node] Add new function overload in 12.10 to fs.rmdir

### LINKED ISSUE #39103: [@types/node] rmdir is missing the recursive option
- [x] I tried using the `@types/node` package and had problems.
- [x] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [x] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).  (Please ask any appropriate questions there).
- [x] [Mention](https://github.com/blog/821-mention-somebody-they-re-notified) the authors (see `Definitions by:` in `index.d.ts`) so they can respond.
  - Authors: @microsoft @jkomyno @alvis @BrunoScheufler @OliverJAsh

Hey there 👋 

Since Node.js 12.10, the `fs.rmdir` function has a new options object which (among other things) allows a `recursive: true` property. This is missing from the current types, for all three versions of the function (synchronous, callback, promise-based).

Could you maybe add this? This would be great 😊

### PR DESCRIPTION
**Pull request by yeldiRium**
Resolves #39103.

In Node.JS 12.10.0 new options were added for [fs.rmdir](https://nodejs.org/api/fs.html#fs_fs_rmdir_path_options_callback), [fs.rmdirSync](https://nodejs.org/api/fs.html#fs_fs_rmdirsync_path_options) and [fs.promises.rmdir](https://nodejs.org/api/fs.html#fs_fspromises_rmdir_path_options).

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [ ] Test the change in your own code. (Compile and run.)
- [ ] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://nodejs.org/api/fs.html (Also individually linked above)
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

### CODE REVIEW

**Comment by yeldiRium** [comment]
The tests (`npm run test`) when run locally fail with the following:

```
...
  /home/yeldir/querbeet/workspace/vendor/DefinitelyTyped/types/rollup-plugin-peer-deps-external: npm install --ignore-scripts --no-shrinkwrap --no-package-lock --no-bin-links --no-save
Error: Error: Command failed: npm install --ignore-scripts --no-shrinkwrap --no-package-lock --no-bin-links --no-save
npm ERR! code ENOENT
npm ERR! syscall lchown
npm ERR! path /home/yeldir/.npm/_locks/staging-4f768f15303abd2d.lock
npm ERR! errno -2
npm ERR! enoent ENOENT: no such file or directory, lchown '/home/yeldir/.npm/_locks/staging-4f768f15303abd2d.lock'
npm ERR! enoent This is related to npm not being able to find a file.
npm ERR! enoent 

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/yeldir/.npm/_logs/2019-10-14T09_38_29_869Z-debug.log

    at ChildProcess.exithandler (child_process.js:295:12)
    at ChildProcess.emit (events.js:209:13)
    at maybeClose (internal/child_process.js:1021:16)
    at Socket.<anonymous> (internal/child_process.js:430:11)
    at Socket.emit (events.js:209:13)
    at Pipe.<anonymous> (net.js:658:12)
npm ERR! code ENOENT
npm ERR! syscall lchown
npm ERR! path /home/yeldir/.npm/_locks/staging-4f768f15303abd2d.lock
npm ERR! errno -2
npm ERR! enoent ENOENT: no such file or directory, lchown '/home/yeldir/.npm/_locks/staging-4f768f15303abd2d.lock'
npm ERR! enoent This is related to npm not being able to find a file.
npm ERR! enoent 

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/yeldir/.npm/_logs/2019-10-14T09_38_29_869Z-debug.log
    at Object.execAndThrowErrors (/home/yeldir/querbeet/workspace/vendor/DefinitelyTyped/node_modules/types-publisher/src/util/util.ts:166:15)
    at /home/yeldir/querbeet/workspace/vendor/DefinitelyTyped/node_modules/types-publisher/src/tester/test-runner.ts:154:24
    at /home/yeldir/querbeet/workspace/vendor/DefinitelyTyped/node_modules/types-publisher/src/util/util.ts:56:28
    at async Promise.all (index 7)
    at Object.nAtATime (/home/yeldir/querbeet/workspace/vendor/DefinitelyTyped/node_modules/types-publisher/src/util/util.ts:51:5)
    at doInstalls (/home/yeldir/querbeet/workspace/vendor/DefinitelyTyped/node_modules/types-publisher/src/tester/test-runner.ts:144:5)
    at Object.runTests [as default] (/home/yeldir/querbeet/workspace/vendor/DefinitelyTyped/node_modules/types-publisher/src/tester/test-runner.ts:75:5)
    at main (/home/yeldir/querbeet/workspace/vendor/DefinitelyTyped/node_modules/types-publisher/src/tester/test.ts:35:5)
npm ERR! Test failed.  See above for more details.
```

That doesn't look like it is related to my changes. Can someone help me out here?

**Comment by SimonSchick** [comment]
Please use the definitions I wrote instead of using hard-to-re-use inline object definitions.
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/38907
I've had some trouble having my PR merged

**Review by trivikr — CHANGES_REQUESTED** [review]

File: types/node/fs.d.ts
Comment (johan13) [reply types/node/fs.d.ts]: No, options are not optional when followed by a required callback argument. The optional arguments have to be at the of the parameter list, don't they?

Comment (johan13) [inline types/node/fs.d.ts]: How about breaking out the options into an interface, just like was done with MakeDirectoryOptions for mkdir?

(I just submitted a duplicate of the PR by mistake: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/39118. I added an interface called RemoveDirectoryOptions.)

Comment (johan13) [reply types/node/fs.d.ts]: To be consistent with the other declarations, we should perhaps add a separate jsdoc comment for each overload?

File: types/node/fs.d.ts
Comment (johan13) [reply types/node/fs.d.ts]: With the changes trivikr suggests we can drop the first declaration.

File: types/node/fs.d.ts
Comment (johan13) [reply types/node/fs.d.ts]: With the changes trivikr suggests we can drop the first declaration.

File: types/node/fs.d.ts
Comment (johan13) [reply types/node/fs.d.ts]: With the changes trivikr suggests we can drop the first declaration.

**Comment by Flarna** [comment]
see also #39116

**Comment by johan13** [comment]
Out of the three duplicate MRs (#39105, #39116 and #39118), my vote is on #39116.

File: types/node/fs.d.ts
Comment (trivikr) [reply types/node/fs.d.ts]: > How about breaking out the options into an interface?

Agree. This was also suggested in https://github.com/DefinitelyTyped/DefinitelyTyped/pull/39105#issuecomment-541788723

Comment (yeldiRium) [reply types/node/fs.d.ts]: Right. That is why i added an overload here (and copied it to all the other places). I adjusted the other places, but this has to stay imo.

Comment (yeldiRium) [reply types/node/fs.d.ts]: I extracted an interface and put it before the first `rmdir` definition. Is it alright the way I did it?

Comment (yeldiRium) [reply types/node/fs.d.ts]: > To be consistent with the other declarations, we should perhaps add a separate jsdoc comment for each overload?

How should I go about that? Should I copy the annotation from above and adjust it?

**Comment by yeldiRium** [comment]
@johan13 I just saw #39118. Can I copy the comments from your pull request? They are very good and I have not added any yet.

**Comment by johan13** [comment]
> @johan13 I just saw #39118. Can I copy the comments from your pull request? They are very good and I have not added any yet.

Of course. Most of my comments are copied from nodejs.com anyway.

File: types/node/fs.d.ts
Comment (johan13) [reply types/node/fs.d.ts]: At the moment the options argument is missing a jsdoc comment.

fs.mkdir recently received a similar `recursive` flag. That code looks like this:
```
/**
 * Asynchronous mkdir(2) - create a directory.
 * @param path A path to a file. If a URL is provided, it must use the `file:` protocol.
 * @param options Either the file mode, or an object optionally specifying the file mode and whether parent folders
 * should be created. If a string is passed, it is parsed as an octal integer. If not specified, defaults to `0o777`.
 */
function mkdir(path: PathLike, options: number | string | MakeDirectoryOptions | undefined | null, callback: (err: NodeJS.ErrnoException | null) => void): void;

/**
 * Asynchronous mkdir(2) - create a directory with a mode of `0o777`.
 * @param path A path to a file. If a URL is provided, it must use the `file:` protocol.
 */
function mkdir(path: PathLike, callback: (err: NodeJS.ErrnoException | null) => void): void;
```

That is, one declaration with the options argument and one without. I was thinking it'd nice to make this declaration consistent with fs.mkdir.

**Comment by yeldiRium** [comment]
I stuck to naming the interfaces `RmdirOptions` (with a lowercase `d`) for consistency with the function name.

**Review by trivikr — APPROVED** [review]

**Review by SimonSchick — CHANGES_REQUESTED** [review]
You forgot to bump up the minimum required version.
Also I'd prefer we merge https://github.com/DefinitelyTyped/DefinitelyTyped/pull/39116 not because it's my PR but because it updates everything for 12.1x and I've been trying to get it out for about a month now 😛

**Comment by johan13** [comment]
I agree about merging #39116 instead of this one. But if for some reason the maintainers are hesitant to merge that big PR, then this PR would be a smaller step in the same direction...

**Comment by SimonSchick** [comment]
I've had bigger PRs merged in the past, I think it's more related to the fact that it's missing approvals and the build is currently failing (for unrelated reasons). My 2 previous PRs were approved but not merged due to package conflicts, I think this will land in a day or 2.

**Comment by yeldiRium** [comment]
Since #39116 has been merged in the meantime, this can be closed :) Thanks all for the feedback and support.

**Pull request closed without merging**
